receiver/snmpreceiver/config.go

```diff
@@ -0,0 +1,616 @@
+}
+
+// SumMetric contains info about the value of the sum metric
+type SumMetric struct {
+	// Aggregation is required and can be cumulative or delta
+	Aggregation string `mapstructure:"aggregation"`
+	// Monotonic is required and can be true or false
+	Monotonic bool `mapstructure:"monotonic"`
+	// ValueType is required can can be either int or float
+	ValueType string `mapstructure:"value_type"`
+}
+
+// ScalarOID holds OID info for a scalar metric as well as any attributes
+// that are attached to it
+type ScalarOID struct {
+	// OID is required and is the scalar OID that is associated with a metric
+	OID string `mapstructure:"oid"`
+	// Attributes is optional and may contain names and values associated with enum
+	// AttributeConfigs to associate with the value of the scalar OID
+	Attributes []Attribute `mapstructure:"attributes"`
+}
+
+// ColumnOID holds OID info for an indexed metric as well as any attributes
+// or resource attributes that are attached to it
+type ColumnOID struct {
+	// OID is required and is the column OID that is associated with a metric
+	OID string `mapstructure:"oid"`
+	// ResourceAttributes is required only if there are no Attributes associated with non enum
+	// AttributeConfigs defined here. Valid values are ResourceAttributeConfig names that will
+	// be used to differentiate the indexed values for the column OID
+	ResourceAttributes []string `mapstructure:"resource_attributes"`
+	// Attributes is required only if there are no ResourceAttributes associated defined here.
+	// Valid values are non enum AttributeConfig names that will be used to differentiate the
+	// indexed values for the column OID
+	Attributes []Attribute `mapstructure:"attributes"`
+}
+
+// Attribute is a connection between a metric configuration and an AttributeConfig
+type Attribute struct {
+	// Name is required and should match the key for an AttributeConfig
+	Name string `mapstructure:"name"`
+	// Value is optional and is only needed for a matched AttributeConfig's with enum value.
+	// Value should match one of the AttributeConfig's enum values in this case
+	Value string `mapstructure:"value"`
+}
+
+// Validate validates the given config, returning an error specifying any issues with the config.
+func (cfg *Config) Validate() error {
+	var combinedErr error
+
+	combinedErr = multierr.Append(combinedErr, validateEndpoint(cfg))
+	combinedErr = multierr.Append(combinedErr, validateVersion(cfg))
+	version := strings.ToUpper(cfg.Version)
+	switch version {
+	case "V3":
+		combinedErr = multierr.Append(combinedErr, validateSecurity(cfg))
+	}
+	combinedErr = multierr.Append(combinedErr, validateMetricConfigs(cfg))
+
+	return combinedErr
+}
+
+// validateEndpoint validates the Endpoint
+func validateEndpoint(cfg *Config) error {
+	if cfg.Endpoint == "" {
+		return errEmptyEndpoint
+	}
+
+	// Add the schema prefix to the endpoint if it doesn't contain one
+	if !strings.Contains(cfg.Endpoint, "://") {
+		cfg.Endpoint = "udp://" + cfg.Endpoint
```

Comment:
Validation of the config should not alter the config.

Have the factory apply this if necessary.

Reply:
@djaglowski I don't know why I missed the fact that I could put this in the createMetricReceiver. I falsely thought that it couldn't be done there. I agree and I'll make this change.